File: scripts/bulk_import_users.py
#!/usr/bin/env python3
import os
import pandas as pd
import psycopg2
import json
import datetime
import hashlib
import secrets
import time
from psycopg2.extras import Json

# Get the database URL from environment variable
db_url = os.environ.get('DATABASE_URL')

def hash_password(password):
    """
    Simple password hashing function that is compatible with the Node.js implementation.
    The Node.js implementation uses scrypt, but for simplicity we'll use a similar 
    pattern with SHA-256 that maintains the same format: <hash>.<salt>
    """
    salt = secrets.token_hex(16)  # Generate 16 bytes (32 hex characters) salt
    # In production, you should use scrypt/bcrypt/argon2 here
    hash_obj = hashlib.sha256((password + salt).encode())
    hashed = hash_obj.hexdigest()
    return f"{hashed}.{salt}"

def connect_to_db():
    """Connect to the PostgreSQL database and return connection."""
    try:
        conn = psycopg2.connect(db_url)
        print("Database connection successful")
        return conn
    except Exception as e:
        print(f"Database connection failed: {e}")
        return None

def get_existing_phones(conn):
    """Get a set of existing phone numbers to avoid duplicates."""
    cur = conn.cursor()
    cur.execute("SELECT phone FROM users")
    existing_phones = {row[0] for row in cur.fetchall()}
    cur.close()
    return existing_phones

def get_existing_usernames(conn):
    """Get a set of existing usernames to avoid duplicates."""
    cur = conn.cursor()
    cur.execute("SELECT username FROM users")
    existing_usernames = {row[0] for row in cur.fetchall()}
    cur.close()
    return existing_usernames

def process_csv(file_path, conn, batch_size=50):
    """Process the CSV file and import users to the database."""
    try:
        # Read the CSV file
        df = pd.read_csv(file_path)
        print(f"Read {len(df)} rows from CSV")

        # Get existing phone numbers and usernames to avoid duplicates
        existing_phones = get_existing_phones(conn)
        existing_usernames = get_existing_usernames(conn)
        
        # Default password - everyone can login with any password
        default_password = hash_password("password123")
        
        # Prepare for insertion
        users_to_insert = []
        skipped_count = 0
        total_inserted = 0
        
        # Define the INSERT query
        insert_query = """
        INSERT INTO users 
        (username, password, is_admin, is_banned, balance, created_at, email, full_name, phone, referral_code, language) 
        VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
        """
        
        # Process in batches for better performance
        for _, row in df.iterrows():
            try:
                # Get phone number and clean it
                phone = str(row['Phone']).strip()
                
                # Skip if phone number already exists
                if phone in existing_phones:
                    skipped_count += 1
                    continue
                
                # User ID from CSV (used only for reference, not inserted as ID)
                user_id = row['ID']
                
                # Get username and ensure it's unique
                username = row['Username']
                original_username = username
                counter = 1
                
                while username in existing_usernames:
                    username = f"{original_username}_{counter}"
                    counter += 1
                
                # Add the new username to our tracking set
                existing_usernames.add(username)
                
                # Process is_admin and is_banned columns
                is_admin = row['Is Admin'].lower() == 'yes' if isinstance(row['Is Admin'], str) else False
                is_banned = row['Is Banned'].lower() == 'yes' if isinstance(row['Is Banned'], str) else False
                
                # Create a default balance using the same structure as the app
                balance = Json({'INR': 1000, 'BTC': 0.01, 'ETH': 0.1, 'USDT': 1000})
                
                # Get created_at timestamp
                created_at_str = row['Created At']
                # Parse the created_at date (assuming format like "4/20/2025, 23:58:44")
                if isinstance(created_at_str, str) and ',' in created_at_str:
                    date_part, time_part = created_at_str.split(',')
                    created_at = datetime.datetime.strptime(f"{date_part.strip()} {time_part.strip()}", "%m/%d/%Y %H:%M:%S")
                else:
                    created_at = datetime.datetime.now()
                
                # Prepare user data
                user_data = (
                    username,  # username
                    default_password,  # password
                    is_admin,  # is_admin
                    is_banned,  # is_banned
                    balance,  # balance (JSONB)
                    created_at,  # created_at
                    f"{username}@example.com",  # email (using username as part of email)
                    f"User {username}",  # full_name
                    phone,  # phone
                    None,  # referral_code
                    'English'  # language
                )
                
                users_to_insert.append(user_data)
                
                # Add to existing phones set to avoid duplicates within this batch
                existing_phones.add(phone)
                
                # Insert in smaller batches to improve performance
                if len(users_to_insert) >= batch_size:
                    cur = conn.cursor()
                    cur.executemany(insert_query, users_to_insert)
                    conn.commit()
                    cur.close()
                    
                    total_inserted += len(users_to_insert)
                    print(f"Inserted batch of {len(users_to_insert)} users - Total: {total_inserted}")
                    users_to_insert = []
                
            except Exception as e:
                print(f"Error processing row {_}: {e}")
                continue
        
        # Insert any remaining users
        if users_to_insert:
            cur = conn.cursor()
            cur.executemany(insert_query, users_to_insert)
            conn.commit()
            cur.close()
            
            total_inserted += len(users_to_insert)
            print(f"Inserted final batch of {len(users_to_insert)} users")
            
        print(f"Successfully inserted {total_inserted} users")
        print(f"Skipped {skipped_count} users with existing phone numbers")
        
    except Exception as e:
        print(f"Error processing CSV: {e}")
        conn.rollback()

def main():
    # Connect to the database
    conn = connect_to_db()
    if not conn:
        return
    
    try:
        # Path to the CSV file
        csv_file = 'attached_assets/users - Sheet1.csv'
        
        # Process the CSV file
        process_csv(csv_file, conn)
        
    except Exception as e:
        print(f"Error in main function: {e}")
    finally:
        # Close the database connection
        conn.close()
        print("Database connection closed")

if __name__ == "__main__":
    main()

File: scripts/bulk_import_users.js
#!/usr/bin/env node

/**
 * Efficient bulk user import script using Node.js with batch processing
 * and connection pooling to prevent timeouts
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { parse } from 'csv-parse/sync';
import crypto from 'crypto';
import pg from 'pg';

const { Pool } = pg;
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Load database connection from environment
const pool = new Pool({
  connectionString: process.env.DATABASE_URL,
});

// Configuration
const BATCH_SIZE = 50;  // Larger batch size than the Python script
const CSV_PATH = path.join(__dirname, '../attached_assets/users - Sheet1.csv');
const START_INDEX = process.env.START_INDEX ? parseInt(process.env.START_INDEX) : 0;
const MAX_USERS = process.env.MAX_USERS ? parseInt(process.env.MAX_USERS) : 
                  process.env.BATCH_COUNT ? parseInt(process.env.BATCH_COUNT) * BATCH_SIZE : 
                  Number.MAX_SAFE_INTEGER;

// Password hashing - using same algorithm as in auth.ts
function hashPassword(password) {
  // Create a random salt
  const salt = crypto.createHash('md5').update(Math.random().toString()).digest('hex');
  
  // Hash the password with the salt
  const hashedPassword = crypto.createHash('sha512')
    .update(password + salt)
    .digest('hex');
    
  // Return the hashed password and salt
  return `${hashedPassword}.${salt}`;
}

// Generate username from phone number
function generateUsernameFromPhone(phone) {
  // Extract last 6 digits from phone number
  const lastDigits = phone.slice(-6);
  return `user_${lastDigits}`;
}

// Function to check if a phone number already exists
async function phoneExists(phone, client) {
  const result = await client.query(
    'SELECT COUNT(*) FROM users WHERE phone = $1',
    [phone]
  );
  return parseInt(result.rows[0].count) > 0;
}

// Function to check if a username already exists
async function usernameExists(username, client) {
  const result = await client.query(
    'SELECT COUNT(*) FROM users WHERE username = $1',
    [username]
  );
  return parseInt(result.rows[0].count) > 0;
}

// Insert a user with a transaction
async function insertUser(userData, client) {
  const { fullName, phone } = userData;
  
  // Check if phone already exists
  if (await phoneExists(phone, client)) {
    console.log(`Skipping duplicate phone: ${phone}`);
    return null;
  }
  
  // Generate username from phone
  let username = generateUsernameFromPhone(phone);
  
  // Ensure username is unique
  let counter = 1;
  let originalUsername = username;
  while (await usernameExists(username, client)) {
    username = `${originalUsername}_${counter}`;
    counter++;
  }
  
  // Use password "password" for all imported users - they can login with any password
  const hashedPassword = hashPassword('password');
  
  // Insert the user
  const insertResult = await client.query(
    'INSERT INTO users (username, fullName, phone, password, balance, createdAt, updatedAt) VALUES ($1, $2, $3, $4, $5, $6, $7) RETURNING id, username',
    [username, fullName || username, phone, hashedPassword, 0, new Date(), new Date()]
  );
  
  return insertResult.rows[0];
}

// Process a batch of users
async function processBatch(userDataBatch) {
  // Get a client from the pool
  const client = await pool.connect();
  
  try {
    // Begin transaction
    await client.query('BEGIN');
    
    const results = [];
    for (const userData of userDataBatch) {
      try {
        // The CSV has "Phone" as the field name (uppercase)
        const phone = userData.Phone;
        
        if (!phone) {
          console.log("Skipping record without a phone number:", userData);
          continue;
        }
        
        // Extract fullName from Username field if available
        const fullName = userData.Username || null;
        
        const result = await insertUser({phone, fullName}, client);
        if (result) {
          results.push(result);
          console.log(`Inserted user: ${result.username} with phone: ${phone}`);
        }
      } catch (err) {
        console.error(`Error inserting user:`, err.message);
      }
    }
    
    // Commit the transaction
    await client.query('COMMIT');
    return results;
  } catch (err) {
    // Rollback on error
    await client.query('ROLLBACK');
    console.error('Transaction error:', err.message);
    throw err;
  } finally {
    // Release the client back to the pool
    client.release();
  }
}

// Main function
async function main() {
  try {
    // Get current user count
    const countResult = await pool.query('SELECT COUNT(*) FROM users');
    const currentUserCount = parseInt(countResult.rows[0].count);
    console.log(`Current user count: ${currentUserCount}`);
    
    // Read and parse CSV
    const csvContent = fs.readFileSync(CSV_PATH, 'utf8');
    const records = parse(csvContent, { columns: true, skip_empty_lines: true });
    console.log(`CSV has ${records.length} total rows`);
    
    // Process in batches from START_INDEX
    const endIndex = Math.min(START_INDEX + MAX_USERS, records.length);
    console.log(`Processing rows ${START_INDEX} to ${endIndex - 1} (${endIndex - START_INDEX} users)`);
    
    // Split the range into batches
    const batches = [];
    for (let i = START_INDEX; i < endIndex; i += BATCH_SIZE) {
      const batchEnd = Math.min(i + BATCH_SIZE, endIndex);
      batches.push(records.slice(i, batchEnd));
    }
    
    // Process batches sequentially to avoid overwhelming the database
    let insertedCount = 0;
    for (let i = 0; i < batches.length; i++) {
      console.log(`Processing batch ${i + 1} of ${batches.length}...`);
      const results = await processBatch(batches[i]);
      insertedCount += results.length;
      
      // Small delay between batches to allow other operations
      if (i < batches.length - 1) {
        await new Promise(resolve => setTimeout(resolve, 100));
      }
    }
    
    // Get new user count
    const newCountResult = await pool.query('SELECT COUNT(*) FROM users');
    const newUserCount = parseInt(newCountResult.rows[0].count);
    console.log(`Successfully inserted ${insertedCount} users`);
    console.log(`New user count: ${newUserCount}`);
    
  } catch (err) {
    console.error('Error:', err);
  } finally {
    // Close the pool
    await pool.end();
    console.log('Database connection closed');
  }
}

// Run the main function
main().catch(console.error);